antarest/matrixstore/matrix_garbage_collector.py

```diff
@@ -0,0 +1,117 @@
+import logging
+import threading
+from os import listdir
+from pathlib import Path
+from typing import Set
+
+from antarest.core.config import Config
+from antarest.core.exceptions import StudyTypeUnsupported
+from antarest.core.jwt import DEFAULT_ADMIN_USER
+from antarest.core.requests import RequestParameters
+from antarest.matrixstore.repository import MatrixDataSetRepository
+from antarest.study.service import StudyService
+from antarest.study.storage.variantstudy.variant_study_service import (
+    VariantStudyService,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class MatrixGarbageCollector:
+    def __init__(self, config: Config, study_service: StudyService):
+        self.saved_matrices_path: Path = config.storage.matrixstore
+        self.managed_studies_path: Path = config.storage.workspaces[
+            "default"
+        ].path
+        self.thread = threading.Thread(target=self._loop, daemon=True)
+        self.study_service: StudyService = study_service
+        self.variant_study_service: VariantStudyService = (
+            study_service.storage_service.variant_study_service
+        )
+        self.dataset_repository: MatrixDataSetRepository = (
+            MatrixDataSetRepository()
+        )
+
+    def _get_saved_matrices(self) -> Set[str]:
+        logging.info("Getting all saved matrices")
+        return {f.split(".")[0] for f in listdir(self.saved_matrices_path)}
+
+    def _get_matrices_used_in_raw_studies(self) -> Set[str]:
+        logger.info("Getting all matrices used in raw studies")
+        return {
+            f.name.split(".")[0]
+            for f in self.managed_studies_path.rglob("*.link")
+        }
+
+    def _get_matrices_used_in_variant_studies(self) -> Set[str]:
+        logger.info("Getting all matrices used in variant studies")
+        params = RequestParameters(user=DEFAULT_ADMIN_USER)
+        list_studies = self.study_service.get_studies_information(
+            summary=True,
+            managed=True,
+            params=params,
+        ).keys()
+        variant_study_commands = []
+        for study_id in list_studies:
+            try:
+                variant_study_commands += (
+                    self.variant_study_service.get_icommands(
+                        variant_study_id=study_id, params=params
+                    )
+                )
+            except StudyTypeUnsupported:
+                pass
```

Comment:
you could just get all commands from the CommandBlock table and then call `variant_study_service.command_factory.to_icommand` on each of them